File: android/src/main/kotlin/com/mytiki/sdk/capture/receipt/capacitor/plugin/js/JSPaymentMethod.kt
/*
 * Copyright (c) TIKI Inc.
 * MIT license. See LICENSE file in root directory.
 */

package com.mytiki.sdk.capture.receipt.capacitor.plugin.js

import com.getcapacitor.JSObject
import com.microblink.core.PaymentMethod
import org.json.JSONObject

/**
 * Represents a Receipt Scanning Plugin Payment Method for TIKI SDK.
 *
 * This class wraps the [PaymentMethod] class provided by MicroBlink
 * and provides a structured representation for payment-related information
 * extracted from a receipt.
 *
 * @param paymentMethod The [PaymentMethod] instance to be wrapped.
 */
class JSPaymentMethod(paymentMethod: PaymentMethod) {
    private val paymentMethod: JSStringType?
    private val cardType: JSStringType?
    private val cardIssuer: JSStringType?
    private val amount: JSFloatType?

    init {
        this.paymentMethod = JSStringType.opt(paymentMethod.paymentMethod())
        cardType = JSStringType.opt(paymentMethod.cardType())
        cardIssuer = JSStringType.opt(paymentMethod.cardIssuer())
        amount = JSFloatType.opt(paymentMethod.amount())
    }

    /**
     * Converts the [JSPaymentMethod] instance to a JSON representation.
     *
     * @return A [JSONObject] containing the payment method information.
     */
    fun toJS(): JSObject =
        JSObject()
            .put("paymentMethod", paymentMethod?.toJS())
            .put("cardType", cardType?.toJS())
            .put("cardIssuer", cardIssuer?.toJS())
            .put("amount", amount?.toJS())

    companion object {
        /**
         * Creates an [JSPaymentMethod] instance from a [PaymentMethod] instance,
         * or returns null if the input is null.
         *
         * @param paymentMethod The [PaymentMethod] instance to be converted.
         * @return An [JSPaymentMethod] instance if [paymentMethod] is not null; otherwise, null.
         */
        fun opt(paymentMethod: PaymentMethod?): JSPaymentMethod? =
            if (paymentMethod != null) JSPaymentMethod(paymentMethod) else null
    }
}

File: ios/Plugin/Plugin/Js/JSPaymentMethod.swift
/*
 * RspPaymentMethod Struct
 * Copyright (c) TIKI Inc.
 * MIT license. See LICENSE file in root directory.
 */

import Foundation
import Capacitor
import BlinkReceipt
import BlinkEReceipt

/**
 Represents a response containing payment method information.

 This struct is used to convey payment method details, such as payment method type, card type, card issuer, and amount.
 */
public struct JSPaymentMethod {
    /// The payment method, if available.
    private let paymentMethod: JSStringType?
    /// The card type, if available.
    private let cardType: JSStringType?
    /// The card issuer, if available.
    private let cardIssuer: JSStringType?
    /// The payment amount, if available.
    private let amount: JSFloatType?
    
    /**
     Initializes an `RspPaymentMethod` struct.

     - Parameters:
        - paymentMethod: The payment method, if available.
        - cardType: The card type, if available.
        - cardIssuer: The card issuer, if available.
        - amount: The payment amount, if available.
     */
    init(method: BRPaymentMethod) {
        self.paymentMethod = JSStringType(stringType: method.method)
        self.cardType = JSStringType(stringType: method.cardType)
        self.cardIssuer = JSStringType(stringType: method.cardIssuer)
        self.amount = JSFloatType(floatType: method.amount)
    }
    
    /**
     Converts the `RspPaymentMethod` struct into a dictionary suitable for use in plugin response data.

     - Returns: A dictionary containing payment method information in a format suitable for a Capacitor plugin call result.
     */
    public func toJSObject() -> JSObject {
        var ret = JSObject()
        ret["paymentMethod"] = paymentMethod?.toJSObject()
        ret["cardType"] = cardType?.toJSObject()
        ret["cardIssuer"] = cardIssuer?.toJSObject()
        ret["amount"] = amount?.toJSObject()
        return ret
    }
}
